File: python/TelegramMonitorCompanion.py
#!/usr/bin/env python3
# TelegramMonitorCompanion.ps1 — portiert nach python
# Quelle: powershell, Projects@Telegram-Monitor:TelegramMonitorCompanion.ps1
# Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

# Telegram Monitor Companion — Starter
#
# Startet den lokalen Monitor im Hintergrund (kein Konsolenfenster), wartet,
# bis der Port wirklich antwortet, und oeffnet die Oberflaeche als eigenes
# Fenster ohne Adressleiste. Laeuft der Monitor schon, wird er nicht erneut
# gestartet — dann wird nur das Fenster geoeffnet.
#
# Aufruf:
#   .\TelegramMonitorCompanion.ps1              starten und oeffnen
#   .\TelegramMonitorCompanion.ps1 -Stop        beenden
#   .\TelegramMonitorCompanion.ps1 -Status      nachsehen, ob er laeuft
#   .\TelegramMonitorCompanion.ps1 -Port 9000   anderer Port
#   .\TelegramMonitorCompanion.ps1 -Console     mit sichtbarem Fenster (Fehlersuche)

import argparse
import os
import sys
import time
import subprocess
import requests
import psutil
from pathlib import Path

def write_step(msg):
    print(f"  {msg}")

def test_monitor(url):
    try:
        response = requests.get(f"{url}/api/status", timeout=2)
        return response.status_code == 200
    except:
        return False

def get_monitor_process(pid_file):
    if not pid_file.exists():
        return None
    try:
        with pid_file.open('r') as f:
            pid = int(f.read().strip())
        return psutil.Process(pid)
    except:
        return None

def find_python():
    """Suche Python im System"""
    candidates = [
        (['py', '-3'], 'py'),
        (['python'], 'python'),
        (['python3'], 'python3')
    ]
    
    for args, cmd in candidates:
        try:
            subprocess.run([cmd, '--version'], 
                         capture_output=True, check=True)
            return args
        except:
            continue
    return None

def main():
    parser = argparse.ArgumentParser(description='Telegram Monitor Companion')
    parser.add_argument('-Port', type=int, default=8765, help='Port number')
    parser.add_argument('-Interval', type=int, default=120, help='Poll interval')
    parser.add_argument('-Stop', action='store_true', help='Stop monitor')
    parser.add_argument('-Status', action='store_true', help='Check status')
    parser.add_argument('-Console', action='store_true', help='Show console window')
    parser.add_argument('-NoBrowser', action='store_true', help='Do not open browser')
    
    args = parser.parse_args()
    
    root = Path(__file__).parent
    pid_file = root / 'data' / 'companion.pid'
    log_file = root / 'data' / 'companion.log'
    url = f"http://127.0.0.1:{args.Port}"
    
    # ---------------------------------------------------------------- beenden ---
    if args.Stop:
        proc = get_monitor_process(pid_file)
        if proc:
            proc.terminate()
            try:
                proc.wait(timeout=5)
            except:
                proc.kill()
            write_step(f"Monitor beendet (PID {proc.pid}).")
        else:
            write_step('Es lief kein Monitor aus diesem Starter.')
        pid_file.unlink(missing_ok=True)
        return
    
    # ----------------------------------------------------------------- Status ---
    if args.Status:
        if test_monitor(url):
            proc = get_monitor_process(pid_file)
            msg = f"Monitor laeuft auf {url}"
            if proc:
                msg += f"  (PID {proc.pid})"
            write_step(msg)
        else:
            write_step(f"Auf {url} antwortet nichts.")
        return
    
    # ------------------------------------------------------------------ Start ---
    print('')
    print('  Telegram Monitor Companion')
    print('  --------------------------')
    
    # Python suchen
    python_cmd = find_python()
    if not python_cmd:
        print('')
        print('  Python wurde nicht gefunden.')
        print('  Herunterladen: https://www.python.org/downloads/')
        print('  Beim Installieren "Add python.exe to PATH" ankreuzen.')
        print('')
        input('  Eingabetaste zum Schliessen')
        sys.exit(1)
    
    write_step(f"Python: {' '.join(python_cmd)}")
    
    if test_monitor(url):
        write_step(f"Monitor laeuft bereits auf {url} — wird nicht erneut gestartet.")
    else:
        # Verzeichnis erstellen
        pid_file.parent.mkdir(parents=True, exist_ok=True)
        
        # Server starten
        server_args = python_cmd + [
            'server.py', 
            '--port', str(args.Port),
            '--poll-interval', str(args.Interval),
            '--no-browser'
        ]
        
        if args.Console:
            proc = subprocess.Popen(
                server_args,
                cwd=str(root)
            )
        else:
            # Ohne Fenster, Ausgabe in die Protokolldatei
            with log_file.open('w') as stdout, log_file.with_suffix('.log.err').open('w') as stderr:
                proc = subprocess.Popen(
                    server_args,
                    cwd=str(root),
                    stdout=stdout,
                    stderr=stderr,
                    creationflags=subprocess.CREATE_NO_WINDOW if sys.platform == 'win32' else 0
                )
        
        # PID speichern
        pid_file.write_text(str(proc.pid))
        write_step(f"Gestartet (PID {proc.pid}), warte auf Antwort ...")
        
        # Auf Antwort warten
        ok = False
        for i in range(40):  # bis zu 20 Sekunden
            time.sleep(0.5)
            if test_monitor(url):
                ok = True
                break
            if proc.poll() is not None:  # Prozess beendet
                break
        
        if not ok:
            print('')
            print('  Der Monitor hat nicht geantwortet.')
            err_file = log_file.with_suffix('.log.err')
            if err_file.exists():
                print('  Letzte Zeilen der Fehlerausgabe:')
                lines = err_file.read_text(encoding='utf-8').splitlines()[-15:]
                for line in lines:
                    print(f"    {line}")
            print('')
            print('  Nochmal mit sichtbarem Fenster:  .\\TelegramMonitorCompanion.ps1 -Console')
            input('  Eingabetaste zum Schliessen')
            sys.exit(1)
        
        write_step('Antwortet.')
    
    if args.NoBrowser:
        write_step(f"Bereit: {url}")
        return
    
    # Als eigenes Fenster oeffnen (App-Modus), sonst normaler Tab
    import webbrowser
    import platform
    
    if platform.system() == 'Windows':
        edge = Path(os.environ.get('ProgramFiles(x86)', '')) / 'Microsoft' / 'Edge' / 'Application' / 'msedge.exe'
        chrome = Path(os.environ.get('ProgramFiles', '')) / 'Google' / 'Chrome' / 'Application' / 'chrome.exe'
        
        if edge.exists():
            subprocess.Popen([str(edge), f"--app={url}"])
            write_step('Als eigenes Fenster geoeffnet (Edge).')
        elif chrome.exists():
            subprocess.Popen([str(chrome), f"--app={url}"])
            write_step('Als eigenes Fenster geoeffnet (Chrome).')
        else:
            webbrowser.open(url)
            write_step('Im Standardbrowser geoeffnet.')
    else:
        webbrowser.open(url)
        write_step('Im Standardbrowser geoeffnet.')
    
    print('')
    print(f"  Laeuft im Hintergrund auf {url}")
    print('  Beenden:  .\\TelegramMonitorCompanion.ps1 -Stop')
    print('')

if __name__ == '__main__':
    main()

File: javascript/TelegramMonitorCompanion.js
#!/usr/bin/env node
// TelegramMonitorCompanion.ps1 — portiert nach javascript
// Quelle: powershell, Projects@Telegram-Monitor:TelegramMonitorCompanion.ps1
// Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

/*
    Telegram Monitor Companion — Starter

    Startet den lokalen Monitor im Hintergrund (kein Konsolenfenster), wartet,
    bis der Port wirklich antwortet, und oeffnet die Oberflaeche als eigenes
    Fenster ohne Adressleiste. Laeuft der Monitor schon, wird er nicht erneut
    gestartet — dann wird nur das Fenster geoeffnet.

    Aufruf:
      node TelegramMonitorCompanion.js              starten und oeffnen
      node TelegramMonitorCompanion.js --stop        beenden
      node TelegramMonitorCompanion.js --status      nachsehen, ob er laeuft
      node TelegramMonitorCompanion.js --port 9000   anderer Port
      node TelegramMonitorCompanion.js --console     mit sichtbarem Fenster (Fehlersuche)
*/

const fs = require('fs');
const path = require('path');
const { spawn, execSync } = require('child_process');
const http = require('http');
const os = require('os');

const argv = require('yargs')
  .option('port', {
    alias: 'p',
    type: 'number',
    default: 8765,
    describe: 'Portnummer'
  })
  .option('interval', {
    alias: 'i',
    type: 'number',
    default: 120,
    describe: 'Poll-Intervall in Sekunden'
  })
  .option('stop', {
    alias: 's',
    type: 'boolean',
    describe: 'Monitor beenden'
  })
  .option('status', {
    alias: 't',
    type: 'boolean',
    describe: 'Status abfragen'
  })
  .option('console', {
    alias: 'c',
    type: 'boolean',
    describe: 'Mit sichtbarem Konsolenfenster starten'
  })
  .option('no-browser', {
    alias: 'n',
    type: 'boolean',
    describe: 'Kein Browserfenster oeffnen'
  })
  .help()
  .argv;

const rootDir = __dirname;
const pidFile = path.join(rootDir, 'data', 'companion.pid');
const logFile = path.join(rootDir, 'data', 'companion.log');
const url = `http://127.0.0.1:${argv.port}`;

function writeStep(msg) {
  console.log(`  ${msg}`);
}

function testMonitor() {
  return new Promise((resolve) => {
    const req = http.get(`${url}/api/status`, { timeout: 2000 }, (res) => {
      resolve(res.statusCode === 200);
    });
    req.on('error', () => resolve(false));
    req.on('timeout', () => {
      req.destroy();
      resolve(false);
    });
  });
}

function getMonitorProcess() {
  if (!fs.existsSync(pidFile)) return null;
  try {
    const pid = parseInt(fs.readFileSync(pidFile, 'utf8').trim(), 10);
    if (isNaN(pid)) return null;
    process.kill(pid, 0); // prüft, ob Prozess existiert
    return { pid };
  } catch (e) {
    return null;
  }
}

// ---------------------------------------------------------------- beenden ---
if (argv.stop) {
  const proc = getMonitorProcess();
  if (proc) {
    try {
      process.kill(proc.pid, 'SIGTERM');
      writeStep(`Monitor beendet (PID ${proc.pid}).`);
    } catch (e) {
      writeStep(`Konnte Monitor nicht beenden (PID ${proc.pid}).`);
    }
  } else {
    writeStep('Es lief kein Monitor aus diesem Starter.');
  }
  if (fs.existsSync(pidFile)) fs.unlinkSync(pidFile);
  process.exit(0);
}

// ----------------------------------------------------------------- Status ---
if (argv.status) {
  testMonitor().then((running) => {
    if (running) {
      const proc = getMonitorProcess();
      writeStep(`Monitor laeuft auf ${url}${proc ? `  (PID ${proc.pid})` : ''}.`);
    } else {
      writeStep(`Auf ${url} antwortet nichts.`);
    }
  });
  return;
}

// ------------------------------------------------------------------ Start ---
console.log('');
console.log('  Telegram Monitor Companion');
console.log('  --------------------------');

// Python suchen
let exe = null;
let pre = [];
const candidates = [
  { e: 'py', a: ['-3'] },
  { e: 'python', a: [] },
  { e: 'python3', a: [] }
];

for (const c of candidates) {
  try {
    execSync(`${c.e} --version`, { stdio: 'ignore' });
    exe = c.e;
    pre = c.a;
    break;
  } catch (e) {
    // continue
  }
}

if (!exe) {
  console.log('');
  console.log('  Python wurde nicht gefunden.');
  console.log('  Herunterladen: https://www.python.org/downloads/');
  console.log('  Beim Installieren "Add python.exe to PATH" ankreuzen.');
  console.log('');
  process.exit(1);
}
writeStep(`Python: ${exe} ${pre.join(' ')}`);

testMonitor().then((running) => {
  if (running) {
    writeStep(`Monitor laeuft bereits auf ${url} — wird nicht erneut gestartet.`);
    openBrowser();
  } else {
    fs.mkdirSync(path.dirname(pidFile), { recursive: true });
    const args = [...pre, 'server.py', '--port', argv.port.toString(), '--poll-interval', argv.interval.toString(), '--no-browser'];

    let proc;
    if (argv.console) {
      proc = spawn(exe, args, { cwd: rootDir, stdio: 'inherit' });
    } else {
      const logStream = fs.createWriteStream(logFile, { flags: 'a' });
      const errStream = fs.createWriteStream(logFile + '.err', { flags: 'a' });
      proc = spawn(exe, args, {
        cwd: rootDir,
        stdio: ['ignore', logStream, errStream],
        detached: true
      });
    }

    fs.writeFileSync(pidFile, proc.pid.toString());
    writeStep(`Gestartet (PID ${proc.pid}), warte auf Antwort ...`);

    let ok = false;
    let attempts = 0;
    const maxAttempts = 40;

    const check = () => {
      attempts++;
      testMonitor().then((isUp) => {
        if (isUp) {
          ok = true;
          writeStep('Antwortet.');
          openBrowser();
        } else if (attempts >= maxAttempts) {
          console.log('');
          console.log('  Der Monitor hat nicht geantwortet.');
          if (fs.existsSync(logFile + '.err')) {
            console.log('  Letzte Zeilen der Fehlerausgabe:');
            const lines = fs.readFileSync(logFile + '.err', 'utf8').split('\n').slice(-15);
            lines.forEach(line => console.log(`    ${line}`));
          }
          console.log('');
          console.log('  Nochmal mit sichtbarem Fenster:  node TelegramMonitorCompanion.js --console');
          process.exit(1);
        } else {
          setTimeout(check, 500);
        }
      });
    };

    proc.on('exit', () => {
      if (!ok) {
        console.log('');
        console.log('  Der Monitor-Prozess ist unerwartet beendet worden.');
        process.exit(1);
      }
    });

    check();
  }
});

function openBrowser() {
  if (argv.noBrowser) {
    writeStep(`Bereit: ${url}`);
    return;
  }

  const edge = path.join(process.env['ProgramFiles(x86)'] || '', 'Microsoft', 'Edge', 'Application', 'msedge.exe');
  const chrome = path.join(process.env['ProgramFiles'] || '', 'Google', 'Chrome', 'Application', 'chrome.exe');

  if (fs.existsSync(edge)) {
    spawn(edge, [`--app=${url}`], { detached: true, stdio: 'ignore' });
    writeStep('Als eigenes Fenster geoeffnet (Edge).');
  } else if (fs.existsSync(chrome)) {
    spawn(chrome, [`--app=${url}`], { detached: true, stdio: 'ignore' });
    writeStep('Als eigenes Fenster geoeffnet (Chrome).');
  } else {
    const start = os.platform() === 'win32' ? 'start' :
                  os.platform() === 'darwin' ? 'open' : 'xdg-open';
    spawn(start, [url], { detached: true, stdio: 'ignore' });
    writeStep('Im Standardbrowser geoeffnet.');
  }

  console.log('');
  console.log(`  Laeuft im Hintergrund auf ${url}`);
  console.log('  Beenden:  node TelegramMonitorCompanion.js --stop');
  console.log('');
}
